core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java

```diff
@@ -675,4 +701,14 @@ private static String resourceAsString(String s) throws IOException {
 				StandardCharsets.UTF_8);
 	}
 
+	@Override
+	protected void finalize() throws Throwable {
+		super.finalize();
+		if (initialized.get()) {
+			logger.error("ShaclSail was finalized (garbage collected) without shutdown() having been called first.");
+		}
+		if (executorService != null) {
+			executorService.shutdownNow();
+		}
+	}
```

Comment:
Also considering finalize(). This helps clean up if the user has forgotten to shutdown the sail. Also logs an error. Maybe this should just be a warning?

Reply:
finalize() is deprecated in java 9, and it's recommended anymore. Phantom references are the way to go now: https://docs.oracle.com/javase/8/docs/api/java/lang/ref/PhantomReference.html

Reply:
Looks fairly complex compared to just using finalize()